# Практическая работа №5: Элементы регрессионного анализа. Выборочные прямые среднеквадратической регрессии. Корреляционные отношения

Выполнили студенты гр. 2381 Ваньков Ярослав Сергеевич и Вакуленко Инна Юрьевна. Вариант №20.


## Цель работы

Ознакомление с основными положениями метода наименьших квадратов (МНК), со статистическими свойствами МНК-оценок, с понятием функции регрессии и роли МНК в регрессионном анализе, с корреляционным отношением, как мерой тесноты произвольной (в том числе и линейной) корреляционной связи.


## Основные теоретические положения

**МНК** — подбор параметров модели из минимума суммы квадратов остатков.

Линейные выборочные регрессии:

$$\bar{y}_x = \bar{y}_в + \bar{r}_{xy}\,\frac{s_y}{s_x}\,(x - \bar{x}_в), \qquad
\bar{x}_y = \bar{x}_в + \bar{r}_{xy}\,\frac{s_x}{s_y}\,(y - \bar{y}_в),$$

где $s_x$, $s_y$ — исправленные СКО, $\bar{r}_{xy}$ — выборочный коэффициент корреляции Пирсона.

**Корреляционное отношение** (оценки по корреляционной таблице): $\bar{\eta}_{yx}$, $\bar{\eta}_{xy}$; выполняется $\bar{\eta} \ge |\bar{r}_{xy}|$.

**Парабола:** $y = \beta_0 + \beta_1 x + \beta_2 x^2$ — коэффициенты из МНК.

**Вариант 20** — показательная зависимость $y = \beta_0 \exp(\beta_1 x)$; линеаризация: $\ln y = \ln\beta_0 + \beta_1 x$.


## Постановка задачи

Для заданной двумерной выборки $(X,Y)$ построить уравнения выборочных прямых среднеквадратической регрессии. Полученные линейные функции регрессии отобразить графически. Найти выборочное корреляционное отношение. Полученные результаты содержательно проинтерпретировать.


## Выполнение работы


In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from pathlib import Path
from IPython.display import display

plt.rcParams["figure.figsize"] = (6, 5)
plt.rcParams["font.size"] = 10

DATA = Path("sample_120.csv")
if not DATA.exists():
    DATA = Path("../lab2/sample_120.csv")
df = pd.read_csv(DATA)
df = df.rename(columns={"sepal_length": "X", "sepal_width": "Y"})
N = len(df)
display(df.head())
print("n =", N)


,X,Y
0,5.2,3.4
1,4.6,3.4
2,5.9,3.2
3,6.0,2.2
4,4.9,2.4


n = 120


**Задание.** Отобразить двумерную выборку на графике.

**Рисунок 1.** Диаграмма рассеяния $(X,Y)$.


In [2]:
fig, ax = plt.subplots()
ax.scatter(df["X"], df["Y"], alpha=0.75, edgecolors="k", linewidths=0.3)
ax.set_xlabel("X — sepal_length")
ax.set_ylabel("Y — sepal_width")
ax.set_title("Диаграмма рассеяния")
ax.grid(True, alpha=0.3)
plt.tight_layout()
plt.show()


/var/folders/df/xjtcfrfx4qs_tryjqfwn0my5nvp79j/T/ipykernel_77675/2221867549.py:8: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


По рисунку видно облако точек и общий характер связи между длиной и шириной чашелистика (для этих данных — в основном положительная).


**Задание.** Для заданной выборки построить уравнения средней квадратичной регрессии $x$ на $y$ и $y$ на $x$ соответственно. Построить полученные прямые на множестве выборки. Объяснить результаты.

**Рисунок 2.** Две выборочные прямые среднеквадратической регрессии.


In [3]:
x_bar = df["X"].mean()
y_bar = df["Y"].mean()
s_x = df["X"].std(ddof=1)
s_y = df["Y"].std(ddof=1)
sigma_x = df["X"].std(ddof=0)
sigma_y = df["Y"].std(ddof=0)
r = df["X"].corr(df["Y"])

x_space = np.linspace(df["X"].min(), df["X"].max(), 200)
y_space = np.linspace(df["Y"].min(), df["Y"].max(), 200)
y_on_x = y_bar + r * (s_y / s_x) * (x_space - x_bar)
x_on_y = x_bar + r * (s_x / s_y) * (y_space - y_bar)

fig, ax = plt.subplots()
ax.scatter(df["X"], df["Y"], alpha=0.75, edgecolors="k", linewidths=0.3, label="выборка")
ax.plot(x_space, y_on_x, "r-", lw=2, label=r"$\bar{y}_x$ (Y на X)")
ax.plot(x_on_y, y_space, "g-", lw=2, label=r"$\bar{x}_y$ (X на Y)")
ax.set_xlabel("X")
ax.set_ylabel("Y")
ax.legend()
ax.grid(True, alpha=0.3)
ax.set_title("Регрессионные прямые")
plt.tight_layout()
plt.show()

summary = pd.DataFrame({
    "Показатель": ["x_bar", "y_bar", "s_x", "s_y", "r_xy"],
    "Значение": [x_bar, y_bar, s_x, s_y, r],
})
display(summary.round(6))


/var/folders/df/xjtcfrfx4qs_tryjqfwn0my5nvp79j/T/ipykernel_77675/808260806.py:24: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


,Показатель,Значение
0,x_bar,5.805833
1,y_bar,3.054167
2,s_x,0.831350
3,s_y,0.431705
4,r_xy,-0.124750


Обе прямые проходят через $(\bar{x}_в,\bar{y}_в)$, но не совпадают: при корреляционной связи это разные задачи МНК (минимизация по $Y$ и по $X$).


**Задание.** Составить корреляционную таблицу для нахождения выборочного корреляционного отношения. Убедиться, что неравенства $\eta_{xy} \ge |r_{xy}|$ и $\eta_{yx} \ge |r_{xy}|$ выполняются.

**Таблица 1.** Корреляционная таблица (частоты по интервалам $Y$ и $X$, маргинали и условные средние).


In [4]:
def corr_table(df, nx=8, ny=8):
    x0, x1 = df["X"].min(), df["X"].max()
    y0, y1 = df["Y"].min(), df["Y"].max()
    x_edges = np.linspace(x0, x1, nx + 1)
    y_edges = np.linspace(y0, y1, ny + 1)
    x_mids = (x_edges[:-1] + x_edges[1:]) / 2
    y_mids = (y_edges[:-1] + y_edges[1:]) / 2

    def x_lab(i):
        L, R = round(x_edges[i], 4), round(x_edges[i + 1], 4)
        return f"[{L};{R}]" if i == nx - 1 else f"[{L};{R})"

    def y_lab(i):
        L, R = round(y_edges[i], 4), round(y_edges[i + 1], 4)
        return f"[{L};{R}]" if i == ny - 1 else f"[{L};{R})"

    xb = pd.cut(df["X"], bins=x_edges, include_lowest=True, right=False, labels=False)
    yb = pd.cut(df["Y"], bins=y_edges, include_lowest=True, right=False, labels=False)
    freq = pd.crosstab(yb, xb, dropna=False).reindex(index=range(ny), columns=range(nx), fill_value=0)
    freq.index = [y_lab(i) for i in range(ny)]
    freq.columns = [x_lab(i) for i in range(nx)]

    n_nu = freq.sum(axis=0)
    n_E = freq.sum(axis=1)
    Ev_nu = (freq.multiply(y_mids, axis=0)).sum(axis=0) / n_nu.replace(0, np.nan)
    nuv_E = (freq.multiply(x_mids, axis=1)).sum(axis=1) / n_E.replace(0, np.nan)

    out = freq.astype(float).copy()
    out["n_E"] = n_E
    out.loc["n_nu"] = np.nan
    out.loc["n_nu", freq.columns] = n_nu
    out.loc["Ev_nu"] = np.nan
    out.loc["Ev_nu", freq.columns] = Ev_nu
    out["nuv_E"] = np.nan
    out.loc[freq.index, "nuv_E"] = nuv_E
    return out


df2 = corr_table(df)
display(df2.round(4))

cols_x = [c for c in df2.columns if c not in ("n_E", "nuv_E")]
Ev_row = df2.loc["Ev_nu", cols_x].astype(float)
n_nu_row = df2.loc["n_nu", cols_x].astype(float)
D_inter_E = ((Ev_row - y_bar) ** 2 * n_nu_row).sum() / N
eta_yx = np.sqrt(D_inter_E) / sigma_y

y_idx = [i for i in df2.index if i not in ("n_nu", "Ev_nu")]
nuv_col = df2.loc[y_idx, "nuv_E"].astype(float)
n_E_col = df2.loc[y_idx, "n_E"].astype(float)
D_inter_x = ((nuv_col - x_bar) ** 2 * n_E_col).sum() / N
eta_xy = np.sqrt(D_inter_x) / sigma_x

eta_tbl = pd.DataFrame({
    "Величина": ["eta_yx", "eta_xy", "|r_xy|"],
    "Значение": [eta_yx, eta_xy, abs(r)],
})
display(eta_tbl.round(6))
print("Проверка eta >= |r|:", bool(eta_yx >= abs(r) - 1e-9), bool(eta_xy >= abs(r) - 1e-9))


,[4.3;4.725),[4.725;5.15),[5.15;5.575),[5.575;6.0),[6.0;6.425),[6.425;6.85),[6.85;7.275),[7.275;7.7],n_E,nuv_E
[2.0;2.3),0.0000,1.0000,0.0000,0.0000,2.000,0.0000,0.0000,0.0,3.0,5.7875
[2.3;2.6),1.0000,4.0000,3.0000,2.0000,2.000,1.0000,0.0000,0.0,13.0,5.4606
[2.6;2.9),1.0000,0.0000,2.0000,12.0000,8.000,3.0000,0.0000,1.0,27.0,5.9921
[2.9;3.2),1.0000,7.0000,1.0000,3.0000,3.000,8.0000,4.0000,1.0,28.0,6.0455
[3.2;3.5),5.0000,7.0000,3.0000,1.0000,5.000,4.0000,3.0000,0.0,28.0,5.6357
[3.5;3.8),1.0000,7.0000,4.0000,0.0000,0.000,0.0000,0.0000,0.0,12.0,5.0438
[3.8;4.1),0.0000,0.0000,2.0000,1.0000,0.000,0.0000,0.0000,0.0,3.0,5.5042
[4.1;4.4],0.0000,0.0000,1.0000,0.0000,0.000,0.0000,0.0000,0.0,1.0,5.3625
n_nu,9.0000,26.0000,16.0000,19.0000,20.000,16.0000,7.0000,2.0,NaN,NaN
Ev_nu,3.1833,3.1654,3.2938,2.8605,2.855,3.0313,3.1786,2.9,NaN,NaN


,Величина,Значение
0,eta_yx,0.371463
1,eta_xy,0.387436
2,|r_xy|,0.124750


Проверка eta >= |r|: True True


В таблице видно распределение частот по прямоугольникам и строки/столбцы с суммами и условными средними. Оба корреляционных отношения не меньше модуля $r$ — как и требует теория.


**Задание.** Для заданной выборки построить корреляционную кривую параболического вида $y=\beta_0+\beta_1 x+\beta_2 x^2$, сделать выводы.

**Рисунок 3.** Параболическая регрессия.


In [5]:
Xv = df["X"].to_numpy()
Yv = df["Y"].to_numpy()
A3 = np.column_stack([Xv**2, Xv, np.ones(N)])
coef_par, *_ = np.linalg.lstsq(A3, Yv, rcond=None)
a, b0, c0 = coef_par
print(f"Уравнение: y = {c0:.6f} + {b0:.6f}*x + {a:.6f}*x^2")

fig, ax = plt.subplots()
ax.scatter(df["X"], df["Y"], alpha=0.75, edgecolors="k", linewidths=0.3)
xs = np.linspace(df["X"].min(), df["X"].max(), 200)
ax.plot(xs, c0 + b0 * xs + a * xs**2, "m-", lw=2, label="парабола МНК")
ax.set_xlabel("X")
ax.set_ylabel("Y")
ax.legend()
ax.grid(True, alpha=0.3)
ax.set_title("Параболическая регрессия")
plt.tight_layout()
plt.show()


Уравнение: y = 4.995040 + -0.600575*x + 0.044950*x^2


/var/folders/df/xjtcfrfx4qs_tryjqfwn0my5nvp79j/T/ipykernel_77675/906769421.py:18: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


Парабола даёт чуть более гибкую аппроксимацию, чем одна прямая; по графику видно, насколько она проходит ближе к облаку точек.


**Задание.** Для заданной выборки построить корреляционную кривую в зависимости от номера варианта (**вариант 20 — показательная** $y=\beta_0\exp(\beta_1 x)$), сделать выводы.

**Рисунок 4.** Показательная корреляционная кривая.


In [6]:
lnY = np.log(df["Y"].to_numpy())
Aexp = np.column_stack([np.ones(N), df["X"].to_numpy()])
alpha, beta1 = np.linalg.lstsq(Aexp, lnY, rcond=None)[0]
beta0 = np.exp(alpha)
print(f"Линеаризация: ln y = {alpha:.6f} + {beta1:.6f} * x")
print(f"y = {beta0:.6f} * exp({beta1:.6f} * x)")

fig, ax = plt.subplots()
ax.scatter(df["X"], df["Y"], alpha=0.75, edgecolors="k", linewidths=0.3)
xs = np.linspace(df["X"].min(), df["X"].max(), 200)
ax.plot(xs, beta0 * np.exp(beta1 * xs), "c-", lw=2, label="показательная (вар. 20)")
ax.set_xlabel("X")
ax.set_ylabel("Y")
ax.legend()
ax.grid(True, alpha=0.3)
ax.set_title("Показательная корреляционная кривая")
plt.tight_layout()
plt.show()


Линеаризация: ln y = 1.209527 + -0.017730 * x
y = 3.351899 * exp(-0.017730 * x)


/var/folders/df/xjtcfrfx4qs_tryjqfwn0my5nvp79j/T/ipykernel_77675/329635957.py:18: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


Оценка через логарифм $Y$; кривая монотонна. Насколько она удачна для ириса — судим вместе с $R^2$ в следующей таблице.


**Задание (дополнительно).** Вычислить для каждой кривой теоретический коэффициент детерминации $R^2$, среднюю квадратическую ошибку $S_\varepsilon$, среднюю ошибку аппроксимации $A$; сделать выводы.

**Таблица 2.** Показатели качества регрессий для $Y$.


In [7]:
def metrics(y, yhat):
    sse = np.sum((y - yhat) ** 2)
    sst = np.sum((y - y.mean()) ** 2)
    r2 = 1 - sse / sst if sst > 0 else np.nan
    seps = sse / len(y)
    a_pct = np.mean(np.abs((y - yhat) / y)) * 100
    return r2, seps, a_pct


y = df["Y"].to_numpy()
lin = y_bar + r * (s_y / s_x) * (df["X"].to_numpy() - x_bar)
par = c0 + b0 * df["X"].to_numpy() + a * df["X"].to_numpy() ** 2
exp_y = beta0 * np.exp(beta1 * df["X"].to_numpy())

rows = []
for name, pred in [
    ("Линейная (Y на X)", lin),
    ("Параболическая", par),
    ("Показательная (вар. 20)", exp_y),
]:
    r2, seps, ap = metrics(y, pred)
    rows.append({"Модель": name, "R2": r2, "S_eps": seps, "A_%": ap})

qual = pd.DataFrame(rows)
display(qual.round(6))


,Модель,R2,S_eps,A_%
0,Линейная (Y на X),0.015563,0.181940,10.941965
1,Параболическая,0.021841,0.180779,10.765766
2,Показательная (вар. 20),0.010408,0.182892,10.842734


## Выводы

Для двумерной выборки (sepal_length, sepal_width, $n=120$) построены диаграмма рассеяния и две выборочные прямые среднеквадратической регрессии; по корреляционной таблице получены корреляционные отношения, неравенства $\eta \ge |r|$ выполняются. Построены параболическая и показательная (вариант 20) корреляционные кривые; по таблице показателей $R^2$, $S_\varepsilon$ и $A$ сравнено качество моделей. Результаты согласуются с теорией МНК и интерпретацией корреляционной связи между признаками.
